pkg/capacity/manager.go

```diff
@@ -134,64 +185,52 @@ func (b *ClusterManager) GetOptimalTargetCluster(ctx context.Context, optimalClu
 		return float64(provisioned1By50)/float64(threshold1) < float64(provisioned2By50)/float64(threshold2)
 	})
 
-	b.lastUsed = optimalTargetClusters[0]
-	return optimalTargetClusters[0], nil
+	b.lastUsed = optimalSpaceProvisioners[0].Spec.ToolchainCluster
+	return b.lastUsed, nil
+}
+
+func matches(spc *toolchainv1alpha1.SpaceProvisionerConfig, predicates []spaceProvisionerConfigPredicate) bool {
+	for _, p := range predicates {
+		if !p(spc) {
+			return false
+		}
+	}
+
+	return true
 }
 
 // getOptimalTargetClusters checks if a preferred target cluster was provided and available from the cluster pool.
 // If the preferred target cluster was not provided or not available, but a list of clusterRoles was provided, then it filters only the available clusters matching all those roles.
 // If no cluster roles were provided then it returns all the available clusters.
 // The function returns a slice with an empty string if not optimal target clusters where found.
-func getOptimalTargetClusters(preferredCluster string, getMemberClusters cluster.GetMemberClustersFunc, clusterRoles []string, conditions ...cluster.Condition) []string {
-	emptyTargetCluster := []string{""}
-	// Automatic cluster selection based on cluster readiness
-	members := getMemberClusters(append(conditions, cluster.Ready)...)
-	if len(members) == 0 {
-		return emptyTargetCluster
-	}
-
-	// extract only names of the available clusters
-	var memberNames []string
-	for _, member := range members {
-		// if cluster-role labels were provided, it will check for matching on the member labels
-		// if no clusterRoles labels are required, then the function will return all member cluster with the `tenant` cluster role label
-		if hasClusterRoles(clusterRoles, member) {
-			memberNames = append(memberNames, member.Name)
+func (b *ClusterManager) getOptimalTargetClusters(ctx context.Context, preferredCluster string, conditions ...spaceProvisionerConfigPredicate) ([]toolchainv1alpha1.SpaceProvisionerConfig, error) {
```

Comment:
I'm wondering if there is any benefit in returning the slice of SpaceProvisionerConfigs instead of strings (as it was before) - it could simplify the code in `GetOptimalTargetCluster` function.
But it's not a big deal, feel free to keep it as it is, just update the comment of the function.

Reply:
one more detail - the last param are not conditions but predicates
```suggestion
func (b *ClusterManager) getOptimalTargetClusters(ctx context.Context, preferredCluster string, predicates ...spaceProvisionerConfigPredicate) ([]toolchainv1alpha1.SpaceProvisionerConfig, error) {
```

Reply:
> I'm wondering if there is any benefit in returning the slice of SpaceProvisionerConfigs instead of strings (as it was before) - it could simplify the code in `GetOptimalTargetCluster` function. But it's not a big deal, feel free to keep it as it is, just update the comment of the function.

Returning the slice of `SpaceProvisionerConfig`s is necessary so that we can later sort the slice in the caller, e.g. here https://github.com/codeready-toolchain/host-operator/pull/991/files#diff-5239dc2fa6b427e2f4b412d7caab33e5ec22a42c63e064e6991906a2e0d97e06R172

Reply:
right, I missed that. Thanks for the explanation :+1: